blocks/main_loop.py

```diff
@@ -0,0 +1,96 @@
+"""The event-based main loop of Blocks."""
+from abc import ABCMeta
+
+from blocks.log import RAMTrainingLog
+from blocks.utils import update_instance
+
+
+class MainLoop(object):
+    """The standard main loop of Blocks.
+
+    In the `MainLoop` a model is trained by a training algorithm using data
+    extracted from a data stream. This process is scrupulously documented
+    in a log object.
+
+    The `MainLoop` itself does very little: only fetching the data from the
+    data stream and feeding it to the algorithm. It expects the extensions
+    to do most of the job. A respective callback of every extension is
+    called at every stage of training. The extensions should communicate
+    between themselves and with the main loop object by means of making
+    records in the log. For instance in order to stop the training
+    procedure an extension can make a record
+    `training_finish_requested=True` in the log. The main loop checks for
+    such a record after every batch and every epoch and terminates when
+    finds it.
+
+    Parameters
+    ----------
+    model : object
+        The model object. It is entirely transparent for the main loop
+        but may be used by extensions.
+    data_stream : instance of :class:`DataStream`.
+        The data stream.
+    algorithm : object
+        The training algorithm.
+    log : instance of :class:`TrainingLog`
+        The log. When not given, a :class:`RAMTrainingLog` is created.
+    extensions : list of :class:`TrainingExtension` instances
+        The training extensions. Will be called in the same order as given
+        here.
+
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model, data_stream, algorithm,
+                 log=None, extensions=None):
+        if not log:
+            log = RAMTrainingLog()
+        if not extensions:
+            extensions = []
+        update_instance(self, locals())
+
+    def _run_extensions(self, method_name, *args):
```

Comment:
Can't this functino (and `check_finish_training`) just be public? I don't see any particular reason to keep them private.